File: 3553.minimum-weighted-subgraph-with-the-required-paths-ii.java
#
# @lc app=leetcode id=3553 lang=java
#
# [3553] Minimum Weighted Subgraph With the Required Paths II
#

# @lc code=start
class Solution {
    public int[] minimumWeight(int[][] edges, int[][] queries) {
        int n = edges.length + 1;
        @SuppressWarnings("unchecked")
        java.util.List<int[]>[] adj = new java.util.List[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new java.util.ArrayList<int[]>();
        }
        for (int[] edge : edges) {
            int u = edge[0], v = edge[1], w = edge[2];
            adj[u].add(new int[]{v, w});
            adj[v].add(new int[]{u, w});
        }
        long[] dep = new long[n];
        int[] lev = new int[n];
        int[][] par = new int[n][18];
        boolean[] visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < 18; j++) {
                par[i][j] = -1;
            }
        }
        java.util.Queue<Integer> q = new java.util.LinkedList<>();
        q.offer(0);
        visited[0] = true;
        dep[0] = 0;
        lev[0] = 0;
        par[0][0] = -1;
        while (!q.isEmpty()) {
            int u = q.poll();
            for (int[] nei : adj[u]) {
                int v = nei[0];
                int w = nei[1];
                if (!visited[v]) {
                    visited[v] = true;
                    dep[v] = dep[u] + w;
                    lev[v] = lev[u] + 1;
                    par[v][0] = u;
                    q.offer(v);
                }
            }
        }
        for (int j = 1; j < 18; j++) {
            for (int i = 0; i < n; i++) {
                int p = par[i][j - 1];
                if (p != -1) {
                    par[i][j] = par[p][j - 1];
                }
            }
        }
        class DistHelper {
            long dist(int u, int v) {
                int x = u, y = v;
                if (lev[x] > lev[y]) {
                    int tmp = x;
                    x = y;
                    y = tmp;
                }
                int delta = lev[y] - lev[x];
                for (int j = 0; j < 18; j++) {
                    if ((delta & (1 << j)) != 0) {
                        y = par[y][j];
                    }
                }
                if (x == y) {
                    return dep[u] + dep[v] - 2L * dep[x];
                }
                for (int j = 17; j >= 0; j--) {
                    if (par[x][j] != -1 && par[y][j] != -1 && par[x][j] != par[y][j]) {
                        x = par[x][j];
                        y = par[y][j];
                    }
                }
                int lca = par[x][0];
                return dep[u] + dep[v] - 2L * dep[lca];
            }
        }
        DistHelper dh = new DistHelper();
        int m = queries.length;
        int[] ans = new int[m];
        for (int i = 0; i < m; i++) {
            int src1 = queries[i][0];
            int src2 = queries[i][1];
            int dest = queries[i][2];
            long d12 = dh.dist(src1, src2);
            long d1d = dh.dist(src1, dest);
            long d2d = dh.dist(src2, dest);
            ans[i] = (int) ((d12 + d1d + d2d) / 2);
        }
        return ans;
    }
}
# @lc code=end

File: 3553.minimum-weighted-subgraph-with-the-required-paths-ii.go
#
# @lc app=leetcode id=3553 lang=golang
#
# [3553] Minimum Weighted Subgraph With the Required Paths II
#

# @lc code=start
func minimumWeight(edges [][]int, queries [][]int) []int {
	n := len(edges) + 1
	const LOG = 18

	adj := make([][][2]int, n)
	for _, edge := range edges {
		u, v, w := edge[0], edge[1], edge[2]
		adj[u] = append(adj[u], [2]int{v, w})
		adj[v] = append(adj[v], [2]int{u, w})
	}

	parent := make([]int, n)
	for i := range parent {
		parent[i] = -1
	}
	parent[0] = 0

	weightedDepth := make([]int, n)
	lev := make([]int, n)

	// BFS for depths and parents
	q := []int{0}
	front := 0
	weightedDepth[0] = 0
	lev[0] = 0
	for front < len(q) {
		u := q[front]
		front++
		for _, e := range adj[u] {
			v, w := e[0], e[1]
			if parent[v] == -1 {
				parent[v] = u
				weightedDepth[v] = weightedDepth[u] + w
				lev[v] = lev[u] + 1
				q = append(q, v)
			}
		}
	}

	// Binary lifting
	par := make([][]int, LOG)
	for k := 0; k < LOG; k++ {
		par[k] = make([]int, n)
		for i := 0; i < n; i++ {
			par[k][i] = -1
		}
	}

	for i := 0; i < n; i++ {
		par[0][i] = parent[i]
	}

	for k := 1; k < LOG; k++ {
		for i := 0; i < n; i++ {
			p := par[k-1][i]
			if p != -1 {
				par[k][i] = par[k-1][p]
			}
		}
	}

	getLCA := func(u, v int) int {
		if lev[u] > lev[v] {
			u, v = v, u
		}
		diff := lev[v] - lev[u]
		for k := 0; k < LOG; k++ {
			if (diff & (1 << k)) != 0 {
				v = par[k][v]
			}
		}
		if u == v {
			return u
		}
		for k := LOG - 1; k >= 0; k-- {
			if par[k][u] != par[k][v] {
				u = par[k][u]
				v = par[k][v]
			}
		}
		return par[0][u]
	}

	getDist := func(u, v int) int {
		lca := getLCA(u, v)
		return weightedDepth[u] + weightedDepth[v] - 2 * weightedDepth[lca]
	}

	ans := make([]int, len(queries))
	for i, query := range queries {
		s1, s2, d := query[0], query[1], query[2]
		d12 := getDist(s1, s2)
		d1d := getDist(s1, d)
		d2d := getDist(s2, d)
		ans[i] = (d12 + d1d + d2d) / 2
	}

	return ans
}
# @lc code=end